be/src/exec/pipeline/scan/connector_scan_operator.cpp

```diff
@@ -152,6 +208,191 @@ connector::ConnectorType ConnectorScanOperator::connector_type() {
+    // produced chunks per 10ms, but cs running time is microsecond unit.
+    double balanced_cs_speed = P.op_pull_chunks * 10000.0 / (P.cs_gen_chunks_time + 1);
+    // consumed chunks per 10ms, but op_running time is nanosecond unit.
+    double op_speed = P.op_pull_chunks * 10000000.0 / (P.op_running_time + 1);
+    // `cs_speed` is speed in this single scan operator.
+    int64_t cs_pull_chunks = P.cs_pull_chunks.load();
+    double cs_speed = cs_pull_chunks * 10000.0 / (P.cs_gen_chunks_time + 1);
+    // chunk source: total io time and running time.
+    // we can see if this is slow device. io_latency in ms unit.
+    int64_t cs_total_io_time = P.cs_total_io_time.load();
+    int64_t cs_total_scan_bytes = P.cs_total_scan_bytes.load();
+    // how many 1MB read.
+    double norm_io_count = (cs_total_scan_bytes * 1.0) / (1024 * 1024);
+    double io_latency = cs_total_io_time * 0.000001 / norm_io_count;
+
+    // adjust routines.
+    auto try_add_io_tasks = [&]() {
+        if (!P.try_add_io_tasks) return true;
+        if (P.last_cs_pull_chunks == cs_pull_chunks) return true;
+        return (cs_speed > (P.last_cs_speed * P.expected_speedup_ratio));
+    };
+    auto do_add_io_tasks = [&]() {
+        P.try_add_io_tasks = true;
+        const int smooth = config::connector_io_tasks_adjust_smooth;
+        P.expected_speedup_ratio =
+                (io_tasks + config::connector_io_tasks_adjust_step + smooth) * 1.0 / (io_tasks + smooth);
+        io_tasks += config::connector_io_tasks_adjust_step;
+    };
+    auto do_sub_io_tasks = [&]() {
+        io_tasks -= config::connector_io_tasks_adjust_step;
+        P.try_add_io_tasks = false;
+        P.try_add_io_tasks_fail_count += 1;
+        if (P.try_add_io_tasks_fail_count >= 4) {
+            P.try_add_io_tasks_fail_count = 0;
+            io_tasks -= config::connector_io_tasks_adjust_step;
+        }
+    };
+
+    auto check_slow_io = [&]() {
+        if (((P.check_slow_io++) % 8) != 0) return;
+        if (io_latency >= 2 * P.slow_io_latency_ms) {
+            io_tasks = std::max(io_tasks, _io_tasks_per_scan_operator / 2);
+        } else if (io_latency >= P.slow_io_latency_ms) {
+            io_tasks = std::max(io_tasks, _io_tasks_per_scan_operator / 4);
+        } else {
+        }
+    };
+
+    // adjust io tasks according to feedback.
+    auto do_adjustment = [&]() {
+        if (balanced_cs_speed > op_speed) {
+            do_sub_io_tasks();
+            return;
+        }
+
+        check_slow_io();
+        if (try_add_io_tasks()) {
+            // if we don't try add io tasks before,
+            // or if we've tried and we get expected speedup ratio.
+            do_add_io_tasks();
+        } else {
+            do_sub_io_tasks();
+        }
+    };
+
+    do_adjustment();
+    io_tasks = std::min(io_tasks, _io_tasks_per_scan_operator);
+    io_tasks = std::max(io_tasks, min_io_tasks);
+
+    auto build_log = [&]() {
```

Comment:
is it expensive?

Reply:
it's only called only when you enable verbose logging, and we disable verbose logging by default.

Reply:
but this functor will be call anyway